pkg/identity/gitlabcom/principal.go

```diff
@@ -0,0 +1,205 @@
+	// https://agent.buildkite.com/.well-known/openid-configuration
+	issuer string
+
+	// The URL of the pipeline, the container of many builds. This will be
+	// set as a human-friendly SubjectAlternativeName URI in the certificate.
+	url string
+
+	// Event that triggered this workflow run. E.g "push", "tag" etc
+	eventName string
+
+	// Pipeline ID
+	pipelineID string
+
+	// Repository building built
+	repository string
+
+	// ID to the source repo
+	repositoryID string
+
+	// Owner of the source repo (mutable)
+	repositoryOwner string
+
+	// ID of the source repo
+	repositoryOwnerID string
+
+	// job ID
+	jobID string
+
+	// Git ref being built
+	ref string
+
+	// Commit SHA being built
+	sha string
+
+	// ID of the runner
+	runnerID int64
+
+	// The type of runner used by the job. May be one of gitlab-hosted or self-hosted.
+	runnerEnvironment string
+}
+
+func JobPrincipalFromIDToken(_ context.Context, token *oidc.IDToken) (identity.Principal, error) {
+	var claims struct {
+		ProjectPath       string `json:"project_path"`
+		ProjectID         string `json:"project_id"`
+		PipelineSource    string `json:"pipeline_source"`
+		PipelineID        string `json:"pipeline_id"`
+		NamespacePath     string `json:"namespace_path"`
+		NamespaceID       string `json:"namespace_id"`
+		JobID             string `json:"job_id"`
+		Ref               string `json:"ref"`
+		RefType           string `json:"ref_type"`
+		Sha               string `json:"sha"`
+		RunnerEnvironment string `json:"runner_environment"`
+		RunnerID          int64  `json:"runner_id"`
+	}
+
+	if err := token.Claims(&claims); err != nil {
+		return nil, err
+	}
+
+	if claims.ProjectPath == "" {
+		return nil, errors.New("missing project_path claim in ID token")
+	}
+
+	if claims.PipelineSource == "" {
+		return nil, errors.New("missing pipeline_source claim in ID token")
+	}
+
+	if claims.PipelineID == "" {
+		return nil, errors.New("missing pipeline_id claim in ID token")
+	}
+
+	if claims.JobID == "" {
+		return nil, errors.New("missing job_id claim in ID token")
+	}
+
+	if claims.Ref == "" {
+		return nil, errors.New("missing ref claim in ID token")
+	}
+
+	if claims.RefType == "" {
+		return nil, errors.New("missing ref_type claim in ID token")
+	}
+
+	if claims.NamespacePath == "" {
+		return nil, errors.New("missing namespace_path claim in ID token")
+	}
+
+	if claims.NamespaceID == "" {
+		return nil, errors.New("missing namespace_id claim in ID token")
+	}
+
+	if claims.ProjectID == "" {
+		return nil, errors.New("missing project_id claim in ID token")
+	}
+
+	if claims.Sha == "" {
+		return nil, errors.New("missing sha claim in ID token")
+	}
+
+	if claims.RunnerEnvironment == "" {
+		return nil, errors.New("missing runner_environment claim in ID token")
+	}
+
+	if claims.RunnerID == 0 {
+		return nil, errors.New("missing runner_id claim in ID token")
+	}
+
+	var ref string
+	switch claims.RefType {
+	case "branch":
```

Comment:
Generally we don't want to mutate any claims, thoughts on removing this? I saw the comment from @wlynch though. If you think this would be good to standardize on, let's first open a PR to discuss requiring this prefix.

Reply:
The problem is without the prefix ref alone is ambiguous - you could have `ref/tags/main` or `ref/heads/v1.0.0`. `ref_type` is what informs what the full-formed ref is. Ideally it would be better if GitLab had a full_ref claim (this way they could also support non branch/tag refs as well), but for now I think this is the best we can do.

cc @marshall007 @Brcrwilliams FYI

Reply:
Given that ref type comes from the token, this looks alright.

Reply:
> Ideally it would be better if GitLab had a full_ref claim 

@wlynch we can certainly do that: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/119075

In the code, we currently refer to this has the `ref_path`. Do you think that makes sense as a user-facing name?

Reply:
That sounds fine. Thanks! 🚀

Reply:
Thanks y'all! So we'll assume that `ref_path` will be equivalent to what's here and will update once it's implemented.

Reply:
Should we wait for the MR on the GitLab side to be merged or we stick with this implementation for now?